// @langchain/langgraph/dist/prebuilt/agent_executor.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createAgentExecutor = void 0;
const runnables_1 = require("@langchain/core/runnables");
const tool_executor_js_1 = require("./tool_executor.cjs");
const state_js_1 = require("../graph/state.cjs");
const index_js_1 = require("../index.cjs");
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function _getAgentState(inputSchema) {
    if (!inputSchema) {
        return {
            input: {
                value: null,
            },
            agentOutcome: {
                value: null,
            },
            steps: {
                value: (x, y) => x.concat(y),
                default: () => [],
            },
        };
    }
    else {
        return inputSchema;
    }
}
function createAgentExecutor({ agentRunnable, tools, inputSchema, }) {
    let toolExecutor;
    if (!Array.isArray(tools)) {
        toolExecutor = tools;
    }
    else {
        toolExecutor = new tool_executor_js_1.ToolExecutor({
            tools,
        });
    }
    const state = _getAgentState(inputSchema);
    // Define logic that will be used to determine which conditional edge to go down
    const shouldContinue = (data) => {
        if (data.agentOutcome && "returnValues" in data.agentOutcome) {
            return "end";
        }
        return "continue";
    };
    const runAgent = async (data) => {
        const agentOutcome = await agentRunnable.invoke(data);
        return {
            agentOutcome,
        };
    };
    const executeTools = async (data) => {
        const agentAction = data.agentOutcome;
        if (!agentAction || "returnValues" in agentAction) {
            throw new Error("Agent has not been run yet");
        }
        const output = await toolExecutor.invoke(agentAction);
        return {
            steps: [[agentAction, output]],
        };
    };
    // Define a new graph
    const workflow = new state_js_1.StateGraph({
        channels: state,
    });
    // Define the two nodes we will cycle between
    workflow.addNode("agent", new runnables_1.RunnableLambda({ func: runAgent }));
    workflow.addNode("action", new runnables_1.RunnableLambda({ func: executeTools }));
    // Set the entrypoint as `agent`
    // This means that this node is the first one called
    workflow.setEntryPoint("agent");
    // We now add a conditional edge
    workflow.addConditionalEdges(
    // First, we define the start node. We use `agent`.
    // This means these are the edges taken after the `agent` node is called.
    "agent", 
    // Next, we pass in the function that will determine which node is called next.
    shouldContinue, 
    // Finally we pass in a mapping.
    // The keys are strings, and the values are other nodes.
    // END is a special node marking that the graph should finish.
    // What will happen is we will call `should_continue`, and then the output of that
    // will be matched against the keys in this mapping.
    // Based on which one it matches, that node will then be called.
    {
        // If `tools`, then we call the tool node.
        continue: "action",
        // Otherwise we finish.
        end: index_js_1.END,
    });
    // We now add a normal edge from `tools` to `agent`.
    // This means that after `tools` is called, `agent` node is called next.
    workflow.addEdge("action", "agent");
    return workflow.compile();
}
exports.createAgentExecutor = createAgentExecutor;
